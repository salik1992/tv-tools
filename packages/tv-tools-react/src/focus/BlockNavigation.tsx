import type { PropsWithChildren } from 'react';
import { FocusContext, useFocusContainer } from '../focus';

function markAsProcessed() {
	return true;
}

/**
 * BlockNavigation is a component that prevents navigation to escape from its children.
 * Useful for Modal, VirtualKeyboard, etc.
 */
export const BlockNavigation = ({ children }: PropsWithChildren) => {
	const { focusContextValue, useOnLeft, useOnUp, useOnDown, useOnRight } =
		useFocusContainer();

	useOnLeft(markAsProcessed);
	useOnRight(markAsProcessed);
	useOnUp(markAsProcessed);
	useOnDown(markAsProcessed);

	return (
		<FocusContext.Provider value={focusContextValue}>
			{children}
		</FocusContext.Provider>
	);
};
