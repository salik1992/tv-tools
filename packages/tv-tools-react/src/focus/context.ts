import { createContext } from 'react';
import { noop } from '@salik1992/tv-tools/utils/noop';
import type { Focus } from './types';

export const FocusContext = createContext<Focus>({
	addChild: noop,
});
