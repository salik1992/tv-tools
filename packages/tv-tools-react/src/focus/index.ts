export { FocusRoot } from './FocusRoot';
export { HorizontalFocus } from './HorizontalFocus';
export { Interactable } from './Interactable';
export { VerticalFocus } from './VerticalFocus';
export { useFocusContainer } from './useFocusContainer';
export { useKeySequenceListener } from './useKeySequenceListener';
