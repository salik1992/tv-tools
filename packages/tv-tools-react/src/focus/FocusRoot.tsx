import type { FocusEvent, KeyboardEvent, PropsWithChildren } from 'react';
import { useCallback } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { focus } from '@salik1992/tv-tools/focus';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager.
 * Use at the root of your application.
 */
export const FocusRoot = ({
	children,
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents,
}: PropsWithChildren<{
	alwaysPreventNavigationalEvents?: boolean;
}>) => {
	const onFocus = useCallback((event: FocusEvent) => {
		focus.handleFocusEvent(event.nativeEvent);
	}, []);

	const keyDownCapture = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keydown', 'capture', event);
	}, []);

	const keyDownBubble = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keydown', 'bubble', event);
		if (
			alwaysPreventNavigationalEvents &&
			isDirectional(event) &&
			document.activeElement?.tagName !== 'INPUT'
		) {
			event.preventDefault();
			event.stopPropagation();
		}
	}, []);

	const keyUpCapture = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keyup', 'capture', event);
	}, []);

	const keyUpBubble = useCallback((event: KeyboardEvent<HTMLElement>) => {
		focus.handleKeyEvent('keyup', 'bubble', event);
		if (
			alwaysPreventNavigationalEvents &&
			isDirectional(event) &&
			document.activeElement?.tagName !== 'INPUT'
		) {
			event.preventDefault();
			event.stopPropagation();
		}
	}, []);

	return (
		<div
			id="focusRoot"
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			{children}
		</div>
	);
};
