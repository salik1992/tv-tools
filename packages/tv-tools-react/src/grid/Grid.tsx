import {
	type MouseEvent,
	type ReactNode,
	type WheelEvent,
	useCallback,
	useEffect,
	useMemo,
	useState,
} from 'react';
import type { ControlEvent } from '@salik1992/tv-tools/focus';
import type {
	GridImplementation,
	GridSetup,
	RenderData,
	RenderDataGroup,
} from '@salik1992/tv-tools/grid';
import { FocusContext, useFocusContainer } from '../focus';
import { useThrottledCallback } from '../utils/useThrottledCallback';

const isFirstGroup = (index: number, elementsPerGroup: number) =>
	Math.floor(index / elementsPerGroup) <= 0;

const isLastGroup = (
	index: number,
	elementsPerGroup: number,
	dataLength: number,
) =>
	Math.floor(index / elementsPerGroup) >=
	Math.floor((dataLength - 1) / elementsPerGroup);

/**
 * Grid component.
 */
export const Grid = <Configuration extends Record<string, unknown>>({
	id,
	Implementation,
	configuration,
	renderGroup,
	orientation = 'vertical',
	focusOnMount = false,
	throttleMs = 300,
	ignoreRtl,
	onDataIndex: outerOnDataIndex,
	scrollToTopWithBack = false,
	moveByOneInBetweenGroups = false,
}: {
	/* The optional ID to use for focus and the grid. */
	id?: string;
	/* The behavior of the grid. */
	Implementation: GridImplementation<GridSetup<Configuration>>;
	/* The configuration of the grid. */
	configuration: Omit<GridSetup<Configuration>, 'id'>;
	/* The function that renders each element. */
	renderGroup: (element: RenderDataGroup) => ReactNode;
	/* Orientation of the grid that will change control keys. Default: vertical */
	orientation?: 'vertical' | 'horizontal';
	/* Whether the horizontal grid should render ltr even in rtl. */
	ignoreRtl?: boolean;
	/* Whether the grid should auto focus when mounted. */
	focusOnMount?: boolean;
	/* The time used for throttling the movement. */
	throttleMs?: number;
	/* The listener for changes in data index. */
	onDataIndex?: (index: number) => void;
	/* Whether the grid should scroll to top when pressing BACK. */
	scrollToTopWithBack?: boolean;
	/* Whether when moving by one element, the focus should move to the next/previous group. */
	moveByOneInBetweenGroups?: boolean;
}) => {
	const {
		container,
		focusContextValue,
		useOnUp,
		useOnDown,
		useOnLeft,
		useOnRight,
		useOnBack,
	} = useFocusContainer(id);
	const [dataIndex, setDataIndex] = useState(0);

	// Grid instance
	const grid = useMemo(
		() =>
			new Implementation(container, {
				id: container.id,
				...configuration,
			}),
		[Implementation, container],
	);

	// Render data from the grid
	const [renderData, setRenderData] = useState(grid.getRenderData());

	// Focusing on mount
	useEffect(() => {
		if (focusOnMount) {
			container.focus({ preventScroll: true });
		}
	}, [focusOnMount]);

	const onRenderData = useCallback(
		(newRenderData: RenderData) => {
			setRenderData(newRenderData);
		},
		[setRenderData],
	);

	const onDataIndex = useCallback(
		(index: number) => {
			setDataIndex(index);
			outerOnDataIndex?.(index);
		},
		[outerOnDataIndex, dataIndex],
	);

	// Backwards movement from keys
	const backward = useThrottledCallback(
		(e?: ControlEvent) => {
			return moveByOneInBetweenGroups ||
				dataIndex % configuration.elementsPerGroup !== 0
				? grid.moveBy(-1, (e?.target as HTMLElement | undefined)?.id)
				: false;
		},
		[grid, renderData, dataIndex],
		{ throttledReturn: true, limitMs: throttleMs },
	);
	const backwardJump = useThrottledCallback(
		(e?: ControlEvent) => {
			return !isFirstGroup(dataIndex, configuration.elementsPerGroup)
				? grid.moveBy(
						-configuration.elementsPerGroup,
						(e?.target as HTMLElement | undefined)?.id,
					)
				: false;
		},
		[grid, renderData, dataIndex],
		{ throttledReturn: true, limitMs: throttleMs },
	);

	// Forwards movement from keys
	const forward = useThrottledCallback(
		(e?: ControlEvent) => {
			return moveByOneInBetweenGroups ||
				dataIndex % configuration.elementsPerGroup !==
					configuration.elementsPerGroup - 1
				? grid.moveBy(1, (e?.target as HTMLElement | undefined)?.id)
				: false;
		},
		[grid, renderData, dataIndex],
		{ throttledReturn: true, limitMs: throttleMs },
	);
	const forwardJump = useThrottledCallback(
		(e?: ControlEvent) => {
			return !isLastGroup(
				dataIndex,
				configuration.elementsPerGroup,
				configuration.dataLength,
			)
				? grid.moveBy(
						configuration.elementsPerGroup,
						(e?.target as HTMLElement | undefined)?.id,
					)
				: false;
		},
		[grid, renderData, dataIndex],
		{ throttledReturn: true, limitMs: throttleMs },
	);

	// Attaching movement functions to key listeners
	if (orientation === 'horizontal') {
		useOnLeft(backwardJump, [backwardJump], { ignoreRtl });
		useOnRight(forwardJump, [forwardJump], { ignoreRtl });
		useOnUp(backward);
		useOnDown(forward);
	} else {
		useOnUp(backwardJump);
		useOnDown(forwardJump);
		useOnLeft(backward, [backward], { ignoreRtl });
		useOnRight(forward, [forward], { ignoreRtl });
	}

	useOnBack(() => {
		if (
			scrollToTopWithBack &&
			!isFirstGroup(dataIndex, configuration.elementsPerGroup)
		) {
			grid.moveTo(0);
			return true;
		}
		return false;
	}, [grid, scrollToTopWithBack, dataIndex]);

	// Scrolling the wheel with pointer wheel.
	const onWheel = useThrottledCallback(
		(e: WheelEvent<HTMLElement>) => {
			const delta = e.deltaX || e.deltaY;
			let processed = false;
			if (delta < 0) {
				processed = grid.moveBy(-configuration.elementsPerGroup);
			} else if (delta > 0) {
				processed = grid.moveBy(configuration.elementsPerGroup);
			}
			if (processed) {
				e.stopPropagation();
			}
		},
		[grid, renderData],
		{ throttledReturn: undefined, limitMs: throttleMs },
	);

	const onMousePrevious = useCallback(
		(e: MouseEvent) => {
			e.preventDefault();
			e.stopPropagation();
			backwardJump();
		},
		[backward],
	);

	const onMouseNext = useCallback(
		(e: MouseEvent) => {
			e.preventDefault();
			e.stopPropagation();
			forwardJump();
		},
		[forward],
	);

	useEffect(() => {
		grid.addEventListener('renderData', onRenderData);
		return () => grid.removeEventListener('renderData', onRenderData);
	}, [grid, onRenderData]);

	useEffect(() => {
		if (typeof onDataIndex === 'function') {
			grid.addEventListener('dataIndex', onDataIndex);
			return () => grid.removeEventListener('dataIndex', onDataIndex);
		}
	}, [grid, onDataIndex]);

	return (
		<FocusContext.Provider value={focusContextValue}>
			<div className="grid" onWheel={onWheel}>
				<div
					className="grid-inner-wrap"
					style={{
						transform: `translate${orientation === 'horizontal' ? 'X' : 'Y'}(-${renderData.gridOffset}px)`,
					}}
				>
					{renderData.groups.map(renderGroup)}
				</div>
				{renderData.previousArrow &&
					!isFirstGroup(
						dataIndex,
						configuration.elementsPerGroup,
					) && (
						<div
							className="mouse-arrow previous"
							onClick={onMousePrevious}
						/>
					)}
				{renderData.nextArrow &&
					!isLastGroup(
						dataIndex,
						configuration.elementsPerGroup,
						configuration.dataLength,
					) && (
						<div
							className="mouse-arrow next"
							onClick={onMouseNext}
						/>
					)}
			</div>
		</FocusContext.Provider>
	);
};
