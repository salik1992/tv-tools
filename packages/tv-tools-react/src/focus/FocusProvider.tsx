import type { ComponentProps, FocusEvent, KeyboardEvent } from 'react';
import { useCallback, useMemo } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { FocusManager } from '@salik1992/tv-tools/focus';
import { noop } from '@salik1992/tv-tools/utils/noop';
import { FocusContext } from './context';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager and
 * it also creates the focus manager instance for the application which is available through
 * the FocusContext or useFocusManager hook.
 * Use at the root of your application.
 * @prop alwaysPreventNavigationalEvents - Use this to prevent TV use their own navigation.
 * @prop ...props - Any other props that you want to pass to the wrapping div element.
 */
export const FocusProvider = ({
	children,
	alwaysPreventNavigationalEvents,
	onFocus: onOuterFocus,
	onKeyDownCapture: onOuterKeyDownCapture,
	onKeyDown: onOuterKeyDown,
	onKeyUpCapture: onOuterKeyUpCapture,
	onKeyUp: onOuterKeyUp,
	...props
}: ComponentProps<'div'> & {
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents?: boolean;
}) => {
	const focusManager = useMemo(() => new FocusManager(), []);

	const onFocus = useCallback(
		(event: FocusEvent<HTMLDivElement>) => {
			focusManager.handleFocusEvent(event.nativeEvent);
			onOuterFocus?.(event);
		},
		[focusManager, onOuterFocus],
	);

	const keyDownCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'capture', event);
			onOuterKeyDownCapture?.(event);
		},
		[focusManager, onOuterKeyDownCapture],
	);

	const keyDownBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyDown?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyDown],
	);

	const keyUpCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'capture', event);
			onOuterKeyUpCapture?.(event);
		},
		[focusManager, onOuterKeyUpCapture],
	);

	const keyUpBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyUp?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyUp],
	);

	const focusContextValue = useMemo(
		() => ({
			focusManager,
			addChild: noop,
		}),
		[focusManager],
	);

	return (
		<div
			{...props}
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			<FocusContext.Provider value={focusContextValue}>
				{children}
			</FocusContext.Provider>
		</div>
	);
};
