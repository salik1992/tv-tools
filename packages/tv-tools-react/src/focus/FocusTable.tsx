import {
	type ComponentProps,
	type ComponentType,
	type PropsWithChildren,
	type ReactNode,
	createContext,
	useCallback,
	useContext,
	useEffect,
	useMemo,
} from 'react';
import type { TableFocusContainer } from '@salik1992/tv-tools/focus';
import { FocusContext } from './context';
import { useTableFocusContainer } from './useTableFocusContainer';

const ChildrenOnlyComponent = ({ children }: PropsWithChildren) => {
	return <>{children}</>;
};

const TableContext = createContext<{
	container?: TableFocusContainer;
	TrComponent: ComponentType<PropsWithChildren<unknown>>;
	TdComponent: ComponentType<PropsWithChildren<unknown>>;
}>({
	TrComponent: ChildrenOnlyComponent,
	TdComponent: ChildrenOnlyComponent,
});

const FocusTd = ({
	rowSpan,
	colSpan,
	children,
	...props
}: PropsWithChildren<{
	colSpan?: number;
	rowSpan?: number;
}>) => {
	const { addChild, ...originalfocusContextValue } = useContext(FocusContext);
	const { TdComponent } = useContext(TableContext);

	const addChildWithSpans = useCallback(
		(childId: string) => {
			addChild(childId, { rowSpan, colSpan });
		},
		[addChild, rowSpan, colSpan],
	);

	const focusContextValue = useMemo(
		() => ({ ...originalfocusContextValue, addChild: addChildWithSpans }),
		[addChildWithSpans, originalfocusContextValue],
	);

	return (
		<FocusContext.Provider value={focusContextValue}>
			<TdComponent {...props}>{children}</TdComponent>
		</FocusContext.Provider>
	);
};

const FocusTr = (props: PropsWithChildren<unknown>) => {
	const { TrComponent, container } = useContext(TableContext);

	useEffect(() => {
		// Needs to run every time to maintain children order.
		container?.addRow();
	});

	return <TrComponent {...props} />;
};

export type FocusTableRenderComponents<
	TrComponent extends ComponentType,
	TdComponent extends ComponentType,
> = {
	FocusTr: (props: ComponentProps<TrComponent>) => ReactNode;
	FocusTd: (
		props: {
			colSpan?: number;
			rowSpan?: number;
		} & ComponentProps<TdComponent>,
	) => ReactNode;
};

export const FocusTable = ({
	id,
	children,
	// @ts-expect-error: TODO
	TableComponent = ChildrenOnlyComponent,
	TrComponent = ChildrenOnlyComponent,
	TdComponent = ChildrenOnlyComponent,
	...tableProps
}: {
	children: (
		components: FocusTableRenderComponents<
			typeof TrComponent,
			typeof TdComponent
		>,
		container: TableFocusContainer,
	) => ReactNode;
	id?: string;
	TableComponent?: ComponentType;
	TrComponent?: ComponentType;
	TdComponent?: ComponentType;
} & ComponentProps<typeof TableComponent>) => {
	const {
		focusContextValue,
		container,
		useOnLeft,
		useOnUp,
		useOnDown,
		useOnRight,
	} = useTableFocusContainer(id);

	useOnLeft(
		(event) =>
			container.moveFocus({ x: -1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnRight(
		(event) =>
			container.moveFocus({ x: 1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnUp(
		(event) =>
			container.moveFocus({ y: -1 }, (event.target as HTMLElement).id),
		[container],
	);

	useOnDown(
		(event) =>
			container.moveFocus({ y: 1 }, (event.target as HTMLElement).id),
		[container],
	);

	const tableContextValue = useMemo(
		() => ({
			container,
			TrComponent,
			TdComponent,
		}),
		[TrComponent, TdComponent],
	);

	return (
		<TableContext.Provider value={tableContextValue}>
			<FocusContext.Provider value={focusContextValue}>
				<TableComponent {...tableProps}>
					{children({ FocusTr, FocusTd }, container)}
				</TableComponent>
			</FocusContext.Provider>
		</TableContext.Provider>
	);
};
