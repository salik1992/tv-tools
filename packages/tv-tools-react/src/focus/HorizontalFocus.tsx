import type { KeyboardEvent, PropsWithChildren } from 'react';
import { useFocusContainer } from './useFocusContainer';
import { FocusContext } from './context';

/**
 * Component for markup of where the focus should navigate horizontally.
 * Direct focusable children will be navigated with left/right arrows in the order
 * they are placed in the DOM structure. This component does not render any
 * element. By default this component changes direction when the document is in the
 * right to left layout.
 *
 * @prop id - optional id if you need to address it manually
 * @prop ignoreRtl - optional override to keep the same direction in ltr/rtl layouts
 *
 * @example
 * ```typescriptreact
 * export const Menu = ({ ... }) => {
 *     // ...
 *     return (
 *          <div id="top-menu">
 *              <HorizontalFocus>
 *                  <MenuItem to="home">Home</MenuItem>
 *                  <MenuItem to="search">Search</MenuItem>
 *                  <MenuItem to="epg">EPG</MenuItem>
 *                  <MenuItem to="profile">My Profile</MenuItem>
 *              </HorizontalFocus>
 *          </div>
 *     )
 * }
 * ```
 * If the MenuItem are focusable (using Interactable component internally),
 * then when they are focused and we press left/right, the focus will move between
 * them in the order home <-> search <-> epg <-> profile.
 */
export const HorizontalFocus = ({
	id,
	ignoreRtl,
	children,
}: PropsWithChildren<{ id?: string; ignoreRtl?: boolean }>) => {
	const { focusContextValue, container, useOnLeft, useOnRight } =
		useFocusContainer(id);

	useOnLeft(
		// @ts-ignore: TODO reversed limitation
		(event: KeyboardEvent<HTMLElement>) => {
			return container.moveFocus(-1, (event.target as HTMLElement).id);
		},
		[container],
		{ ignoreRtl },
	);

	useOnRight(
		// @ts-ignore: TODO reversed limitation
		(event: KeyboardEvent<HTMLElement>) => {
			return container.moveFocus(1, (event.target as HTMLElement).id);
		},
		[container],
		{ ignoreRtl },
	);

	return (
		<FocusContext.Provider value={focusContextValue}>
			{children}
		</FocusContext.Provider>
	);
};
