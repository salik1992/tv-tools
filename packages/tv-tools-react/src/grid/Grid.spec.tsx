import { act, render } from '@testing-library/react';
import { DOWN, LEFT, RIGHT, UP } from '@salik1992/tv-tools/control';
import { focus } from '@salik1992/tv-tools/focus';
import { BasicGrid } from '@salik1992/tv-tools/grid/BasicGrid';
import { Performance } from '@salik1992/tv-tools/utils/Performance';
import { FocusRoot, Interactable } from '../focus';
import { Grid } from './Grid';

const focusSpy = jest.spyOn(focus, 'focus');
const throttleMs = 300;
jest.useFakeTimers();

describe('Grid', () => {
	beforeEach(() => {
		focusSpy.mockClear();
	});

	it('should render the grid', () => {
		const { container, unmount } = render(
			<FocusRoot>
				<Grid
					id="grid"
					Implementation={BasicGrid}
					throttleMs={throttleMs}
					configuration={{
						performance: Performance.BASIC,
						dataLength: 25,
						visibleGroups: 5,
						elementsPerGroup: 5,
						config: {
							navigatableGroups: 3,
							scrolling: {
								first: 50,
								other: 100,
							},
						},
					}}
					renderGroup={({ id, offset, elements }) => (
						<div
							className="group"
							id={id}
							key={id}
							data-data-offset={offset}
						>
							{elements.map(({ id, dataIndex, onFocus }) => (
								<Interactable
									id={id}
									key={id}
									data-data-index={dataIndex}
									onFocus={onFocus}
									onPress={jest.fn()}
								/>
							))}
						</div>
					)}
					focusOnMount
				/>
			</FocusRoot>,
		);
		expect(container.innerHTML).toMatchSnapshot();
		expect(focusSpy).toHaveBeenCalledWith('grid-g0-e0', {
			preventScroll: true,
		});
		unmount();
	});

	it('should move with keyboard, mouse wheel and virtual arrows', () => {
		const { container, unmount } = render(
			<FocusRoot>
				<Grid
					id="grid"
					Implementation={BasicGrid}
					throttleMs={throttleMs}
					configuration={{
						performance: Performance.BASIC,
						dataLength: 25,
						visibleGroups: 5,
						elementsPerGroup: 5,
						config: {
							navigatableGroups: 3,
							scrolling: {
								first: 50,
								other: 100,
							},
						},
					}}
					renderGroup={({ id, offset, elements }) => (
						<div
							className="group"
							id={id}
							key={id}
							data-data-offset={offset}
						>
							{elements.map(({ id, dataIndex, onFocus }) => (
								<Interactable
									id={id}
									key={id}
									data-data-index={dataIndex}
									onFocus={onFocus}
									onPress={jest.fn()}
								/>
							))}
						</div>
					)}
					focusOnMount
				/>
			</FocusRoot>,
		);

		// RIGHT
		act(() => {
			container
				.querySelector('#grid-g0-e0')
				?.dispatchEvent(RIGHT.toKeyboardEvent('keydown'));
		});
		expect(focusSpy).toHaveBeenCalledWith('grid-g0-e1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// DOWN
		act(() => {
			container
				.querySelector('#grid-g0-e1')
				?.dispatchEvent(DOWN.toKeyboardEvent('keydown'));
		});
		expect(focusSpy).toHaveBeenCalledWith('grid-g1-e1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// Wheel forward
		act(() =>
			container
				.querySelector('#grid-g1-e1')
				?.dispatchEvent(
					new WheelEvent('wheel', { deltaY: 1, bubbles: true }),
				),
		);
		expect(focusSpy).toHaveBeenCalledWith('grid-g2-e1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// Next arrow
		act(() =>
			container
				.querySelector('.mouse-arrow.next')
				?.dispatchEvent(new MouseEvent('click', { bubbles: true })),
		);
		expect(focusSpy).toHaveBeenCalledWith('grid-g2-e1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// Previous arrow
		act(() =>
			container
				.querySelector('.mouse-arrow.previous')
				?.dispatchEvent(new MouseEvent('click', { bubbles: true })),
		);
		expect(focusSpy).toHaveBeenCalledWith('grid-g1-e1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// Wheel backward
		act(() =>
			container
				.querySelector('#grid-g1-e1')
				?.dispatchEvent(
					new WheelEvent('wheel', { deltaY: -1, bubbles: true }),
				),
		);
		expect(focusSpy).toHaveBeenCalledWith('grid-g1-e1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// UP
		act(() => {
			container
				.querySelector('#grid-g1-e1')
				?.dispatchEvent(UP.toKeyboardEvent('keydown'));
		});
		expect(focusSpy).toHaveBeenCalledWith('grid-g0-e1', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		// LEFT
		act(() => {
			container
				.querySelector('#grid-g0-e1')
				?.dispatchEvent(LEFT.toKeyboardEvent('keydown'));
		});
		expect(focusSpy).toHaveBeenCalledWith('grid-g0-e0', {
			preventScroll: true,
		});
		act(() => jest.advanceTimersByTime(throttleMs));

		unmount();
	});
});
