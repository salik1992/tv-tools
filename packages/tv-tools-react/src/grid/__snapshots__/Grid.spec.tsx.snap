// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should render the grid 1`] = `"<div id="focusRoot"><div class="grid"><div class="grid-inner-wrap" style="transform: translateY(-0px);"><div class="group" id="grid-g0" data-data-offset="0"><div data-data-index="0" id="grid-g0-e0" tabindex="0"></div><div data-data-index="1" id="grid-g0-e1" tabindex="0"></div><div data-data-index="2" id="grid-g0-e2" tabindex="0"></div><div data-data-index="3" id="grid-g0-e3" tabindex="0"></div><div data-data-index="4" id="grid-g0-e4" tabindex="0"></div></div><div class="group" id="grid-g1" data-data-offset="0"><div data-data-index="5" id="grid-g1-e0" tabindex="0"></div><div data-data-index="6" id="grid-g1-e1" tabindex="0"></div><div data-data-index="7" id="grid-g1-e2" tabindex="0"></div><div data-data-index="8" id="grid-g1-e3" tabindex="0"></div><div data-data-index="9" id="grid-g1-e4" tabindex="0"></div></div><div class="group" id="grid-g2" data-data-offset="0"><div data-data-index="10" id="grid-g2-e0" tabindex="0"></div><div data-data-index="11" id="grid-g2-e1" tabindex="0"></div><div data-data-index="12" id="grid-g2-e2" tabindex="0"></div><div data-data-index="13" id="grid-g2-e3" tabindex="0"></div><div data-data-index="14" id="grid-g2-e4" tabindex="0"></div></div><div class="group" id="grid-g3" data-data-offset="0"><div data-data-index="15" id="grid-g3-e0" tabindex="0"></div><div data-data-index="16" id="grid-g3-e1" tabindex="0"></div><div data-data-index="17" id="grid-g3-e2" tabindex="0"></div><div data-data-index="18" id="grid-g3-e3" tabindex="0"></div><div data-data-index="19" id="grid-g3-e4" tabindex="0"></div></div><div class="group" id="grid-g4" data-data-offset="0"><div data-data-index="20" id="grid-g4-e0" tabindex="0"></div><div data-data-index="21" id="grid-g4-e1" tabindex="0"></div><div data-data-index="22" id="grid-g4-e2" tabindex="0"></div><div data-data-index="23" id="grid-g4-e3" tabindex="0"></div><div data-data-index="24" id="grid-g4-e4" tabindex="0"></div></div></div><div class="mouse-arrow next"></div></div></div>"`;
